api/websocket/messagebuffer/messagebuffer.go

```diff
@@ -1,39 +1,218 @@
+	size  int
+}
+
+func New() *Buffer {
+	return &Buffer{
+		queue: make(map[string][]*queue, 0),
+	}
+}
+
+func getParent(index int) int {
+	if index <= 0 {
+		return -1
+	}
+
+	return (index - 1) / 2
+}
+
+func getChildren(index int) (int, int) {
+	tree := index * 2
+
+	return tree + 1, tree + 2
+}
+
+func (q *queue) isAfter(node *queue) bool {
+	a := q.item
+	b := node.item
+
+	if a.creationTime.Before(b.creationTime) {
+		return false
+	}
+	if a.creationTime == b.creationTime {
+		return false
+	}
+
+	return true
+}
+
+func (q *queue) isAfterOrEqual(node *queue) bool {
+	a := q.item
+	b := node.item
+
+	if a.creationTime.Before(b.creationTime) {
+		return false
+	}
+	if a.creationTime == b.creationTime {
+		return true
+	}
+
+	return true
+}
+
+func (m *Buffer) Purge() {
+	var clients []string
+
+	m.mux.Lock()
+	for clientIDStr := range m.queue {
+		clients = append(clients, clientIDStr)
+	}
+	m.mux.Unlock()
+
+	for _, clientIDStr := range clients {
+		relays := m.PopMessages(clientIDStr)
+
+		for _, relay := range relays {
+			m.put(clientIDStr, relay)
+		}
 	}
 }
 
-// AddMessage adds a message to message buffer
-func (messageBuffer *MessageBuffer) AddMessage(clientID []byte, msg *pb.Relay) {
-	clientIDStr := hex.EncodeToString(clientID)
-	messageBuffer.Lock()
-	defer messageBuffer.Unlock()
-	messageBuffer.buffer[clientIDStr] = append(messageBuffer.buffer[clientIDStr], msg)
+func (m *Buffer) AddMessage(clientID string, relay *pb.Relay) error {
+	if relay.MaxHoldingSeconds <= 0 {
+		return nil
+	}
+
+	m.Purge()
+
+	if m.size >= config.Parameters.MaxClientMessageBufferSize {
+		return fmt.Errorf("Max buffer size reached.")
+	}
+
+	return m.put(clientID, relay)
 }
 
-// PopMessages reads and clears all messages of a client
-func (messageBuffer *MessageBuffer) PopMessages(clientID []byte) []*pb.Relay {
-	clientIDStr := hex.EncodeToString(clientID)
-	messageBuffer.Lock()
-	defer messageBuffer.Unlock()
-	messages := messageBuffer.buffer[clientIDStr]
-	messageBuffer.buffer[clientIDStr] = nil
+func (m *Buffer) PopMessages(clientID string) []*pb.Relay {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+
+	var messages []*pb.Relay
+
+	s := 0
+	item, _ := m.get(clientID)
+
+	for item != nil {
+		if item.expiryTime.Before(time.Now()) {
+			continue
+		}
+
+		messages = append(messages, item.relay)
+		s += len(item.relay.Payload)
+
+		item, _ = m.get(clientID)
+	}
+
+	m.size -= s
+
 	return messages
 }
+
+func (m *Buffer) put(clientID string, relay *pb.Relay) error {
+	m.mux.Lock()
+	defer m.mux.Unlock()
+
+	q := m.queue[clientID]
+	next := len(q)
+	now := time.Now()
+	node := &queue{
+		parent: m,
+		item: &Item{
+			creationTime: now,
+			expiryTime:   now.Add(time.Second * time.Duration(relay.MaxHoldingSeconds)),
+			relay:        relay,
+		},
+	}
+
+	m.queue[clientID] = append(m.queue[clientID], node)
+
+	par := getParent(next)
+	cur := next
+
+	for (par >= 0) && (q[par].isAfter(node)) {
+		m.queue[clientID][par], m.queue[clientID][cur] = q[cur], q[par]
```

Comment:
This seems not right. len(q) is not increased after append (but len(m.queue[clientID]) is), so you are accessing q[len(q)]